docs/scaling/SYSTEM.md

```diff
@@ -0,0 +1,190 @@
+
+### ServerlessServices (SKS)
+
+**API:** `serverlessservices.networking.internal.knative.dev`
+
+An abstraction on top of Kubernetes Services that controls the data flow and
+especially the switch between having the activator as a buffer on the data path
+(for example when scaled to zero) or routing to the application instances
+directly (when scaled above zero).
+
+This is achieved by creating two Kubernetes services for each revision: A
+**public **service and a **private **service.
+
+The private service is a standard Kubernetes service. Its selector is set to
+point to the application instances deployed and as the deployment grows and
+shrinks, the list of available IPs will do so too.
+
+The public service is an “unmanaged” Kubernetes service. It has no selector and
+therefore does not automatically get endpoint management as the private service
+does. The public service’s endpoints are managed by the SKS reconciler directly.
+
+The SKS has two modes: **Proxy** and **Serve**.
+
+In mode “Serve”, the public service’s endpoints are exactly the same as those of
+the private service. All traffic will flow to the revision’s pods.
+
+In mode Proxy, the public service’s endpoints are the addresses of all
+activators in the system. All traffic will flow to the activators.
+
+ServerlessServices are automatically created from PodAutoscalers by default.
+
+## Data flow examples
+
+<!-- Editable versions of the diagrams below are in the Knative shared drive in Scaling/images -->
+
+### Scaling up and down (steady state)
+
+![scale-up-down](images/scale-up-down.png)
+
+At steady state, the autoscaler is constantly scraping the currently active
+revision pods to adjust the scale of the revision constantly. As requests flow
+into the system, the scraped values will change and the autoscaler will instruct
+the revision's deployment to adhere to a given scale.
+
+The SKS keeps track of the changes to the deployment's size through the private
+service. It updates the public service accordingly.
+
+### Scaling to zero
+
+![scale-to-0](images/scale-to-0.png)
+
+A revision scales to zero once there are no more requests in the system anymore.
+All scrapes from the autoscaler to revision pods return 0 concurrency and the
+activator reports the same.
+
+Before actually removing the last pod of the revision the system makes sure that
+the activator is in the path and routable. The autoscaler, who decided to
+scale-to-zero in the first place, instructs the SKS to use **Proxy** mode, so
+all traffic is directed at the activators. The SKS’ public service is now probed
+until it has been ensured to return responses from the activator. Once that is
+the case and if a grace-period (configurable via _scale-to-zero-grace-period_)
+has already passed, the last pod of the revision is removed and the revision has
+successfully scaled to zero.
+
+### Scaling from zero
+
+![scale-from-0](images/scale-from-0.png)
```

Comment:
- 5.1 should be `proxy`?
-

Reply:
Hmm, I don't think so? It only stays in proxy if TBC is enabled, doesn't it?